### PR TITLE
Add more ways to read scheduled enclave

### PR DESCRIPTION
**Pull request by Traf333**
### Context

this PR adds ~2 ways~ a way to read scheduled enclave list
1. ~via binary:`./bitacross-worker scheduled-enclave`~
2. via RPC: `echo '{"jsonrpc":"2.0","method":"state_getScheduledEnclaveList","id":1}' | websocat -n1 -k -B 99999999 wss://localhost:2000`

### CODE REVIEW

File: bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs
Comment (felixfaisal) [inline bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: This is a `RwLock`, you have to first obtain the `read` lock and then try to get the value. Reference - https://github.com/litentry/litentry-parachain/blob/d893864ee3dbded7d05e9c7921936a30c58841c5/bitacross-worker/litentry/core/scheduled-enclave/src/lib.rs#L51C44-L52C1

Comment (felixfaisal) [reply bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: Use this function instead `get_current_mrenclave()`, reference - https://github.com/litentry/litentry-parachain/blob/d893864ee3dbded7d05e9c7921936a30c58841c5/bitacross-worker/litentry/core/scheduled-enclave/src/io.rs#L134

File: bitacross-worker/service/src/main_impl.rs
Comment (Traf333) [inline bitacross-worker/service/src/main_impl.rs]: it seems like getting result via RPC works fine, but if I want to obtain data via binary it returns empty result and sometimes fails to run. Do you know why it might happen? @kziemianek @felixfaisal 

on the screenshot it shows when triggering binary several times, some executions are failing and at the end I've got data via RPC and it works fine. Might it mean that `GLOBAL_SCHEDULED_ENCLAVE.registry` is every time different?


image

Comment (Kailai-Wang) [reply bitacross-worker/service/src/main_impl.rs]: The error implies a race condition.

`Component factory has not been set` means you try to access the global variable before it's initialized (most likely GLOBAL_SCHEDULED_ENCLAVE in this case)

Comment (Kailai-Wang) [reply bitacross-worker/service/src/main_impl.rs]: It's not necessarily your code's problem - it can be "you simply run the test binary too early"

Comment (Kailai-Wang) [reply bitacross-worker/service/src/main_impl.rs]: In general I recommend adding more lifecycle events representating the worker state (not in this PR ofc) - currently the only thing we have is `EnclaveAdded`

File: bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs
Comment (silva-fj) [inline bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: should this be behind the "development" feature flag ?

Comment (kziemianek) [inline bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: I suggest naming it `state_getScheduledEnclaveList`

File: bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs
Comment (kziemianek) [inline bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: You should embed it with `RpcReturnValue` to be consistent with most of methods.

**Review by kziemianek — CHANGES_REQUESTED** [review]

**Comment by kziemianek** [comment]
Do you plan adding this method to `tee-worker`?

File: bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs
Comment (Kailai-Wang) [reply bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: It should be OK for production too as it's meant to be public information

Comment (silva-fj) [reply bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: I see, if this is not secret info why do we seal it?

Comment (Kailai-Wang) [reply bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: To prevent it from being modified (intentionally or unintentionally) by worker operators. So the public should have read access but no write access.

I mean even if you can't read it, you should be able to infer the content just by checking the parachain tx and tee-worker code : )

Comment (silva-fj) [reply bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: Got it 👍🏼

**Comment by Traf333** [comment]
> Do you plan adding this method to `tee-worker`?
@kziemianek I'd say yes and have only one concern, how would it work with multi-worker setup. Does `scheduled_enclave_sealed.bin` synchronises for each worker once it's run?

**Comment by silva-fj** [comment]
> > Do you plan adding this method to `tee-worker`?
> > @kziemianek I'd say yes and have only one concern, how would it work with multi-worker setup. Does `scheduled_enclave_sealed.bin` synchronises for each worker once it's run?

It will once we merge https://github.com/litentry/litentry-parachain/pull/2628

File: bitacross-worker/service/src/main_impl.rs
Comment (Traf333) [reply bitacross-worker/service/src/main_impl.rs]: It seems that this issue `Component factory has not been set` happens with other commands as well. Do we need need to create an issue for that or not worth it? as it seems really low impact

**Comment by Traf333** [comment]
at the end decided to keep only one method to obtain scheduled enclave list as through the binary it seems unstable

**Review by silva-fj — APPROVED** [review]

File: bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs
Comment (kziemianek) [inline bitacross-worker/enclave-runtime/src/rpc/worker_api_direct.rs]: This results in something like this:
`Ok({0: [177, 155, 22, 22, 112, 239, 100, 252, 239, 19, 10, 78, 170, 5, 181, 229, 13, 234, 219, 91, 29, 160, 23, 147, 133, 219, 42, 14, 151, 253, 19, 134]})`

1. You should return only value contained in `Result` not `Result` itself.
2. In my opinion value should be serialized directly, not formated to string then serialized.
3. Can we have an integration test for this ?

**Review by kziemianek — CHANGES_REQUESTED** [review]

**Comment by Traf333** [comment]
Included integration tests which also addresses next linear issue [P-645 : Add test case for scheduled-enclave setting](https://linear.app/litentry/issue/P-645/add-test-case-for-scheduled-enclave-setting) where tests covered:
- `state_getScheduledEnclave` 
- couple of invalid and valid scenarios for `state_setScheduledEnclave`

**Review by kziemianek — APPROVED** [review]
Nicely done

File: tee-worker/ts-tests/integration-tests/scheduled_enclave.test.ts
Comment (silva-fj) [inline tee-worker/ts-tests/integration-tests/scheduled_enclave.test.ts:12]: why do we define the type and then cast it to `any`?

**Review by silva-fj — APPROVED** [review]

File: tee-worker/ts-tests/integration-tests/scheduled_enclave.test.ts
Comment (Traf333) [reply tee-worker/ts-tests/integration-tests/scheduled_enclave.test.ts:12]: good question 🤔 this one was copy-paste from another test. I just created [chore task](https://linear.app/litentry/issue/P-686/worker-ts-tests-cleanup) to cleanup our tests, as I've also seen some fixme's were copied from one test to another

**Pull request merged**
